LoTS: Introduce the assembler-first flow

arthur791004 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/83390

## Proposed Changes

* Introduce the assembler-first flow to move the Assembler before domains & plans.
* The feature is controlled by the `themes/assembler-first` flag
* I'll create a follow-up PR to display the launchpad for this flow after updating your homepage in the Site Editor

### User without any un-launched sites

https://github.com/Automattic/wp-calypso/assets/13596067/775d5170-edb9-4a65-8002-66fc018f3bb2

### Use with un-launched sites

https://github.com/Automattic/wp-calypso/assets/13596067/037bb09d-f748-41d4-8ebe-aafba236543a

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to logged-out TS
* Select the Assembler CTA
* Sign up or Sign in
* If your account has an un-launched site, you will see “New or existing site”. Pick an existing site or create a new site.
* After the site selection or the site creation, you will see the Assembler immediately
* Finish the Assembler and ensure you land on the Site Editor

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

arthur791004 committed:
Introduce the assembler-first flow

arthur791004 commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nav-sidebar/src/components/nav-sidebar/index.tsx:
```diff
@@ -204,8 +195,15 @@ function WpcomBlockEditorNavSidebar() {
 				// Waiting for jsx-a11y version bump to support aria-description attribute
-				// eslint-disable-next-line jsx-a11y/aria-props
-				aria-description={ dialogDescription }
+				// 		aria-description={ postType.slug === 'page'
```
I'm not sure why I cannot disable this rule, `jsx-a11y/role-supports-aria-props`

arthur791004 committed:
Refactor: Move the setActiveChecklist function to the skipLaunchpad function

arthur791004 commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nav-sidebar/src/components/nav-sidebar/index.tsx:
```diff
@@ -204,8 +195,17 @@ function WpcomBlockEditorNavSidebar() {
 				// Waiting for jsx-a11y version bump to support aria-description attribute
-				// eslint-disable-next-line jsx-a11y/aria-props
-				aria-description={ dialogDescription }
+				// aria-description={
```
I'm not sure why I cannot disable this rule, `jsx-a11y/role-supports-aria-props`

taipeicoder replied:
@arthur791004 does this mean that you made changes to this file, the linter threw an error about `aria-description` so that's why it's now commented out?

arthur791004 replied:
YES!

arthur791004 commented on client/landing/stepper/declarative-flow/free.ts:
```diff
@@ -91,9 +90,9 @@ const free: Flow = {
+					return navigate( 'site-creation-step' );
 
-				case 'siteCreationStep':
+				case 'site-creation-step':
```
Rename the slug of the step

fushar replied:
But why?

arthur791004 replied:
That's the slug used everywhere else.

miksansegundo replied:
By the way, I remember we have been asked a few times why we are using camelCase in step names and params like in `/setup/with-theme-assembler/patternAssembler?siteSlug={ SITE }`. See the most recent time in p1700321967885689/1700040293.699609-slack-C048CUFRGFQ


We could start by adding support for kebab-case on the steps and params we use in the assembler without removing support for camelCase. WDYT? cc: @Automattic/lego

miksansegundo replied:
Tracked in https://github.com/Automattic/wp-calypso/issues/84846

fushar commented:
Some issues that I found:

1. It looks like we're always using Creatio 2 as the source site, even though I selected an existing site. Is this expected?

1. The progress bar texts seem way too large than usual (?) also, sometimes I see progress bar, sometimes I see three-dots...

    <img width="607" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1525580/78e2d7a5-66ea-4158-b443-8bb2baae0c3e">

    <img width="494" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1525580/b85930fb-a2ae-41bb-8656-5afa17e2fb95">


1. Back button is missing from all screens:
    
    <img width="299" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1525580/34e733fd-2a08-446e-9778-4b81a2cb6bd2">

1. There is `isNewSite=undefined` when selecting an existing site:

    <img width="155" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1525580/607cb4e2-1a33-4082-a54a-b2a32069afc7">

1. When clicking "Skip for now" in the Launchpad, we're getting redirected to a blank screen:

    <img width="906" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1525580/0b16ee85-e397-488d-8c9e-4b0932dcdf91">

    <img width="301" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1525580/0d437312-dc0b-457f-b3a6-e2bba46d063c">

arthur791004 committed:
Fix the Back button is missing

arthur791004 committed:
Avoid adding isNewSite qs when the value is falsy

arthur791004 committed:
Fix the destination after skipping Launchpad

arthur791004 commented:
> It looks like we're always using Creatio 2 as the source site, even though I selected an existing site. Is this expected?

What do you mean here? I selected an existing site, and it seems to work well 🤔

For example, the site title of my existing site is `Hello`

![image](https://github.com/Automattic/wp-calypso/assets/13596067/dc2ab47f-82be-4deb-ab28-fd482ef1485a)

> The progress bar texts seem way too large than usual (?) also, sometimes I see progress bar, sometimes I see three-dots...

It's the existing behavior... We can address this later.

> Back button is missing from all screens:

Ooops... It's the side effect from https://github.com/Automattic/wp-calypso/pull/84368. Fixed now.

> There is isNewSite=undefined when selecting an existing site:

Fixed. I guess this one is the reason for the first issue you mentioned 🤔

> When clicking "Skip for now" in the Launchpad, we're getting redirected to a blank screen:

Fixed because it uses `siteId ?? siteSlug` bu the value of `siteId` may be `0` 😓

taipeicoder commented:
> If your account has an un-launched site, you will see “New or existing site”. Pick an existing site or create a new site.

@arthur791004 do you know why it only looks at un-launched sites? Does this mean that if the user has a launched site, then it won't see this screen?

taipeicoder commented:
I'm still seeing the upsell nudges and the upsell screen in the assembler-first flow:

![Screenshot 2023-12-05 at 8 57 59 AM](https://github.com/Automattic/wp-calypso/assets/797888/4b2ff16f-ed5c-4ec0-9264-95d8f7869581)

I think we can address this in a follow-up PR since this one is already pretty sizable 🙂

arthur791004 commented:
> do you know why it only looks at un-launched sites? Does this mean that if the user has a launched site, then it won't see this screen?

This flow is for the LoTS now so it would be better to let the user choose any un-launched site. That is, the flow should create a new site when people land on it. However, we want to avoid creating too many sites unexpectedly so we allow the user to select the sites that have just been created (un-launched sites).

> Not sure if intended, but using an existing site, the Assembler has a Back button that takes them to the LiTS and no Launchpad is displayed.

Yes, I noticed this and I will address it in the follow-up PR 🙂

> I'm still seeing the upsell nudges and the upsell screen in the assembler-first flow:

YES. We want to keep it and decide whether to leave it as it is in the internal CfT.

taipeicoder commented on client/landing/stepper/declarative-flow/assembler-first-flow.ts:
```diff
@@ -0,0 +1,264 @@
+				design_type: 'assembler',
+			} );
+
+			setIntent( SiteIntent.WithThemeAssembler );
```
@arthur791004 do you think we need to differentiate the intent `with-theme-assembler` and `assembler-first`?

arthur791004 replied:
Oh yes, it should be `assembler-first` 🙈

autumnfjeld commented:
> This flow is for the LoTS now so it would be better to let the user choose any un-launched site. That is, the flow should create a new site when people land on it. However, we want to avoid creating too many sites unexpectedly so we allow the user to select the sites that have just been created (un-launched sites).

Curios where the concern of "too many sites" comes from.  Is this reported? I assume this is a very edge case.  Most people 
, almost all, starting from LoTS will be signing up with a new account.

autumnfjeld commented:
Just noting that in the CfT we should point out that we were undecided on this screen and we'd like discussion in the CfT comments.   The more I think about it, it seems like it could be ok to leave in the flow.   🤔 

![image](https://github.com/Automattic/wp-calypso/assets/5665959/70cb4bbe-7cca-456d-b854-722ef3c21cfd)

autumnfjeld commented:
I just tested the flow and the Launchpad did not appear after I exited the editor.  Instead I just see My Home. 

I will test again to be sure.

arthur791004 committed:
Fix the intent

arthur791004 commented:
> Curios where the concern of "too many sites" comes from. Is this reported? I assume this is a very edge case. Most people
, almost all, starting from LoTS will be signing up with a new account.

@autumnfjeld It's the same as “design-first” or “start-writing-first“ flow. This kind of flow will create a new site automatically when the user enters the flow. If the user refreshes the page or opens the same page multiple times, then they will have lots of new sites.

arthur791004 commented:
> I just tested the flow on a brand new account and the Launchpad did not appear after I exited the editor. Instead I just see My Home.

It's expected as followed: 

> I'll create a follow-up PR to display the launchpad for this flow after updating your homepage in the Site Editor

autumnfjeld commented:
I am seeing the domains page after user account:

https://github.com/Automattic/wp-calypso/assets/5665959/6c16945a-09ef-4817-a345-61b2d6744433

arthur791004 commented:
> Using the Calypso live link, I see the domains page after user account:

Do you add the flag, `?flags=themes/assembler-first` to the URL?

taipeicoder reviewed (approved):
Tested:

- ✅ Works well when selecting the new site option.
- ✅ Works well when selecting the existing site option.
- ✅ Verified that the `newsletter` flow works as before.
- ✅ Verified that the `with-theme` flow works as before.

autumnfjeld commented:
LOL. No. Oops  I followed these instructions:
![image](https://github.com/Automattic/wp-calypso/assets/5665959/10ace33f-c222-436f-b7f1-73276c11f3b4)

fushar commented on client/landing/stepper/declarative-flow/assembler-first-flow.ts:
```diff
@@ -0,0 +1,264 @@
+		// We also need to support both query param and path suffix localized urls
+		// depending on where the user is coming from.
+		const useLocaleSlug = useLocale();
+		// Query param support can be removed after dotcom-forge/issues/2960 and 2961 are closed.
```
I see that both issues are closed now 😄

arthur791004 replied:
Let's leave it as it is now and clean it (ALL usages) up in the follow-up PR 😆
